fix(tianmu): fix STR_TO_DATE() Function does not work well. (#975)

lujiashun opened the pull request:
<!--

Thank you for contributing to StoneDB!

PR Title Format: <type>(<scope>): description to this pr (#issue_id)
e.g.
fix(util): fix sth..... (#1)

-->

## Summary about this PR
<!--

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
e.g.:
Issue: close #1

-->

Issue Number: ref #975

1 add MicroSecond logic;
2 allow "00-00-00 11:11:11" datime format to follow innodb behavior;
3 refactor code,change CT to ColumnType;

## Tests Check List
<!-- At least one of next options must be included. -->

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog
<!-- At least one of next options must be included. -->

- [x] New Feature
- [ ] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation
<!-- At least one of next options must be included. -->

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

RingsC commented on storage/tianmu/types/value_parser4txt.cpp:
```diff
@@ -662,7 +662,8 @@ common::ErrorCode ValueParserForText::ParseDateTimeOrTimestamp(const BString &ti
 
   if (tianmu_err_code != common::ErrorCode::OUT_OF_RANGE && buflen != 0)
-    year = TianmuDateTime::ToCorrectYear((uint)year, at);
+    if (year != kTianmuDateSpec.Year())
```
```
if (a)
if (b)
xxx
```
can be combine into 
```
if (a && b)
  xxx
```

lujiashun replied:
Fixed it.

RingsC commented on storage/tianmu/types/value_parser4txt.cpp:
```diff
@@ -743,10 +746,22 @@ common::ErrorCode ValueParserForText::ParseDateTimeOrTimestamp(const BString &ti
+    tianmu_err_code = system::EatUInt(buf, buflen, microsecond);
+    if (!common::IsError(tianmu_err_code)) {
+      if (!TianmuDateTime::CanBeMicroSecond(microsecond)) {
+        rcv = TianmuDateTime((short)year, (short)month, (short)day, (short)hour, (short)minute, (short)second,
```
using c++ style type casting.

lujiashun replied:
Fixed it.

RingsC commented on storage/tianmu/types/tianmu_data_types.h:
```diff
@@ -301,6 +301,8 @@ class TianmuDateTime : public ValueBasic<TianmuDateTime> {
   TianmuDateTime(int64_t dt, common::CT at);
```
`CT` can be `ColumnType`, it can be in an other PR.

lujiashun replied:
Fixed it.

RingsC reviewed (approved):
LGTM

Commit:
fix(tianmu): fix STR_TO_DATE() Function does not work well. (#975)

[summary]
1 add MicroSecond logic;
2 allow "00-00-00 11:11:11" datime format to follow innodb behavior;
3 refactor code,change CT to ColumnType;

duanfuxiang0 reviewed (approved):
LGTM

mergify[bot] committed:
Merge branch 'stonedb-5.7-dev' into fix-975-stonedb5.7
